Option to toggle fields in songs, albums & artists

aldrinjenson opened the pull request:
Closes /related to #846 

- Add option to toggle song fields/attributes in songslist
- Add new SongFieldsMenu component for toggling Material-ui Menu
- Add constant file for storing constants in order to prevent immutability bugs
- Add fieldsToDisplay state to redux store and hence the user's preference will be safely saved in LocalStorage

Note that the mobile view will not be affected and the default initial view will be prevailing.

https://user-images.githubusercontent.com/53407417/112766491-ac174180-902f-11eb-9774-810de1afbbfa.mp4

aldrinjenson commented:
P.S. The alignment issue of the hearts/starrs have been fixed in a recent pr.

certuna commented:
This PR isn't really related to https://github.com/navidrome/navidrome/issues/846 : 
- that issue is about the Album Details view, your PR is for the Song List view
- that issue deals with auto-hiding redundant information, your PR is about manually enabling/disabling columns

That said, it's a great feature to add, and this could also be added to the Artist List and Album List views!

aldrinjenson commented:
Okay cool. Tho, is it better to add toggle options to artist and album views as a separate pr from this one or to add on together here?

deluan commented:
This looks great, thanks!! I'll take a deep look tomorrow.

> Tho, is it better to add toggle options to artist and album views as a separate pr from this one or to add on together here?

It is better to make separated PRs to facilitate the review, but wait till I review this one, to avoid potential mistakes in the other PRs.

deluan reviewed (commented):
Hey, this is great, the only issue I see is that the code becomes very messy, and currently the logic you implemented is not very reusable. What do you think of making the code more generic, by "borrowing" the idea and api from [React-Admin's Enterprise Edition](https://marmelab.com/ra-enterprise/modules/ra-preferences#selectcolumnsbutton-store-datagrid-columns-in-preferences). Do you think it is possible?

aldrinjenson commented:
Yeah, When the code was first written, I had just the songs list in mind and hence the re-usability part wasn't given much attention. In retrospect I should've realized that hardcoding the source values and assigning a Boolean condition with a wrapper function to check the displaying of each field is too much repeating and could look messy as the fields may increase in the future. 
Let me try and see if it's possible to create our own custom `useSelectedColumns` hook (like the one mentioned in the React-Admin docs) or maybe even a higher order component to wrap over the child columns with an omit prop (fields which should not be displayed based on the user's preference) so as to make the whole flow more reusable and simple.

aldrinjenson commented:
Hi, I've added menu for toggling fields in ArtistLists, SongsLists and AlbumList view.
I've made a `useSelectedColumns` hook similar to the one in [ra-preferences] (https://marmelab.com/ra-enterprise/modules/ra-preferences#selectcolumnsbutton-store-datagrid-columns-in-preferences) with a very simple(I think) API.
A new `ToggleFieldsMenu` component has been made which can be directly used in any new lists which would need toggling. Only the resource name need to be passed as prop.

deluan commented on ui/src/common/ToggleFieldsMenu.js:
```diff
@@ -0,0 +1,77 @@
+
+const ITEM_HEIGHT = 70
+
+export default function ToggleFieldsMenu({ resource }) {
```
Maybe set propTypes for this, marking `resource` as Required?

aldrinjenson replied:
Yeah. sure

deluan commented on ui/src/common/ToggleFieldsMenu.js:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  return (
+    <div style={{ position: 'relative', top: '-0.5em' }}>
```
Please use `makeStyles`

aldrinjenson replied:
Okay.

deluan commented on ui/src/common/useSelectedFields.js:
```diff
@@ -0,0 +1,43 @@
+import { useDispatch, useSelector } from 'react-redux'
+import { setToggleableFields } from '../actions'
+
+const useSelectedFields = ({ resource, columns }) => {
```
It would be awesome to have unit tests for this :)

aldrinjenson replied:
I'll have to dig in a bit at testing :). I'll look at the examples and will write tests. Also I'm thinking of modifying the logic a bit more inorder to make it more compatible with future use cases. Will let you know soon! :crossed_fingers:

deluan commented on ui/src/album/AlbumListActions.js:
```diff
@@ -57,6 +59,9 @@ const AlbumListActions = ({
         </Button>
       </ButtonGroup>
+      {isDesktop && !albumView.grid ? (
+        <ToggleFieldsMenu resource="album" />
```
Not sure if this is the best place for Albums. It makes the other buttons move.... Can't think of a better solution, though...

https://user-images.githubusercontent.com/331353/114949165-67a4e780-9e1e-11eb-9d34-b1fe124cbddb.mov

aldrinjenson replied:
Ahh yes. The other option was to keep the menu icon there on both views. Maybe we could ask the opinion of others as well.

deluan reviewed (changes requested):
Well done, looks great. Please see my comments below. Also, we are still missing it in the `AlbumSongs.js` and `PlaylistSongs.js` grids

aldrinjenson commented on ui/src/common/useSelectedFields.js:
```diff
@@ -0,0 +1,64 @@
+      dispatch(setOmittedFields({ [resource]: omittedColumns }))
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
```
If in the future ND, uses some libraries for rehydrating redux store after new releases like [redux-persist](https://www.npmjs.com/package/redux-persist), then this useEffect could be skipped then.

deluan commented:
Hey @aldrinjenson , did you find a better solution for the ToggleFieldsMenu in the AlbumGrid? I'd love to merge this, but the current solution makes the UI very... weird.

Maybe we could leave the icon in place, but disable it when the view mode is grid. What do you think?

aldrinjenson commented:
Hey @deluan,
> Maybe we could leave the icon in place, but disable it when the view mode is grid. What do you think?

I'm a bit worried that leaving the icon in place and disabling it could cause some confusion maybe. And since the element does not serve any purpose in gridView, I think a better solution would be to change it's visibility to hidden in grid view.
This will ensure that the when toggling between list and grid views, there won't be any jumping for the rest of buttons. 
Personally I think the empty space here is more reasonable than having a menu which does not work.
Here's a demo:

https://user-images.githubusercontent.com/53407417/117008960-de8c1c80-ad08-11eb-9eda-5f50f6bc76c7.mp4

P.S. I tried moving the menu to right-end of the grid header itself by positioning absolutely: but this turned out to have a lot of ugly issues with react-admin.

deluan commented:
Another option is to allow the ToggleFieldsMenu to receive a "TopBar" component. If it was preset, it would be rendered above the column selection, like in [this example](https://marmelab.com/ra-enterprise-demo/#/customers?filter={}) from the original RA Enterprise Edition:

https://user-images.githubusercontent.com/331353/117014014-63057d80-acbe-11eb-8773-d07c4c292bb5.mov

aldrinjenson commented:
Hey @deluan , I've moved the view toggler to the menu like in the example.

![image](https://user-images.githubusercontent.com/53407417/117204984-a242e400-ae0e-11eb-9c67-2ba98e7f7af9.png)
![image](https://user-images.githubusercontent.com/53407417/117205569-6bb99900-ae0f-11eb-8915-1cbd13964545.png)

However, there is one particular edge case yet to be resolved which occurs the very first time Navidrome is loaded (when the localStorage is empty of preferences). The preferences get stored when the useSelectedFields hook is called for the first time. This call occurs in the table view as it is here that the columns are actually toggled. As grid view is the default first view in Navidrome, at the very first launch, the menuItems will be empty.



> It would also be great to have a way to specify the default fields to be toggled on when there are no preferences saved.


Yeah. I've added a new prop called "defaultOff" to the useSelectedFields hook which accepts a list of field-keys that is to be toggled off by default. An example can be seen in songsList.js(line: 118) and albumSongs.js(line 124).

deluan commented on ui/src/album/AlbumActions.js:
```diff
@@ -28,6 +29,7 @@ const AlbumActions = ({
   const dispatch = useDispatch()
   const translate = useTranslate()
   const isDesktop = useMediaQuery((theme) => theme.breakpoints.up('md'))
+  const isSmall = useMediaQuery((theme) => theme.breakpoints.up('sm'))
```
This would be "Not Small", right? Small would be `down('sm')`. This causes confusion when reading things like
`{isSmall && <ToggleFieldsMenu resource="albumSong" />}` when it really means "Show the menu if it is NOT small".

deluan commented on ui/src/album/AlbumActions.js:
```diff
@@ -51,41 +53,52 @@ const AlbumActions = ({
-        onClick={handlePlay}
-        label={translate('resources.album.actions.playAll')}
+      <div
+        style={{
```
This should go into a `useClasses` hook. Please avoid using inline `style` tag

deluan commented on ui/src/playlist/PlaylistActions.js:
```diff
@@ -94,47 +96,58 @@ const PlaylistActions = ({ className, ids, data, record, ...rest }) => {
-        onClick={handlePlay}
-        label={translate('resources.album.actions.playAll')}
+      <div
+        style={{
```
This should go into a useClasses hook. Please avoid using inline style tag

deluan commented on ui/src/album/AlbumListActions.js:
```diff
@@ -37,26 +92,9 @@ const AlbumListActions = ({
-          <ViewHeadlineIcon fontSize="inherit" />
-        </Button>
-      </ButtonGroup>
+      {isSmall && (
```
In the case of Small screens, we need to put back the original grid/list selector, as the ToggleFieldsMenu is hidden

deluan commented on ui/src/album/AlbumSongs.js:
```diff
@@ -87,6 +88,44 @@ const AlbumSongs = (props) => {
+    resource: 'albumSong',
+    columns: toggleableFields,
+    omittedColumns: ['title'],
+    defaultOff: ['rating'],
```
`rating` should not be off by default, but `bpm` should.

deluan commented on ui/src/common/ToggleFieldsMenu.js:
```diff
@@ -0,0 +1,112 @@
+
+ToggleFieldsMenu.propTypes = {
+  resource: PropTypes.string.isRequired,
+  TopBarComponent: PropTypes.elementType,
```
Properties names should be `camelCase` (starting with lowercase): `topbarComponent`

deluan commented on ui/src/common/useSelectedFields.js:
```diff
@@ -0,0 +1,65 @@
+    if (!omittedFields) {
+      dispatch(setOmittedFields({ [resource]: omittedColumns }))
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Why disabling the linter? This error is a real concern. Please add the dependencies

deluan commented on ui/src/common/useSelectedFields.js:
```diff
@@ -0,0 +1,65 @@
+      if (omittedFields.length !== omitted.length)
+        dispatch(setOmittedFields({ [resource]: omitted }))
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Why disabling the linter? This error is a real concern. Please add the dependencies

deluan commented on ui/src/song/SongList.js:
```diff
@@ -77,6 +78,47 @@ const SongList = (props) => {
+  const columns = useSelectedFields({
+    resource: 'song',
+    columns: toggleableFields,
+    defaultOff: ['trackNumber', 'rating'],
```
`rating` should be on by default. `bpm` should be off

deluan commented on ui/src/song/SongListActions.js:
```diff
@@ -20,6 +22,7 @@ export const SongListActions = ({
   ids,
   ...rest
 }) => {
+  const isSmall = useMediaQuery((theme) => theme.breakpoints.up('sm'))
```
This `isSmall` happens in a lot of places, see my comment about it being "inverted". It should be renamed or changed in all places

deluan reviewed (changes requested):
It looks really good. Just some minor changes and we should be good to merge

deluan commented:
Hey @aldrinjenson, I did a small change and implemented a different workaround for the edge case of the first time loading the album list. Your workaround was forcing the default album view to be "list", but the default must be "grid".

Anyway, thanks very much for your hard work on this.
